os/kill.go

```diff
@@ -61,27 +74,123 @@ func getAllChildren(proc *process.Process) []*process.Process {
 // wrapWithStopAndCont wraps original process tree signal sending with SIGSTOP and
 // SIGCONT to prevent processes from forking during termination, so we will not
 // have orphaned processes after.
-func wrapWithStopAndCont(signal syscall.Signal, pgids []int) error {
+func wrapWithStopAndCont(signal syscall.Signal, pgids []int) []syscall.Signal {
 	signals := []syscall.Signal{syscall.SIGSTOP, signal}
 	if signal != syscall.SIGKILL { // no point in sending any signal after SIGKILL
 		signals = append(signals, syscall.SIGCONT)
 	}
+	return signals
+}
 
-	for _, currentSignal := range signals {
-		if err := sendSignalToProcessGroups(currentSignal, pgids); err != nil {
-			return err
+func sendSignalsToProcessGroups(signals []syscall.Signal, pgids []int) error {
+	for _, signal := range signals {
+		for _, pgid := range pgids {
+			log.Infof("Sending signal %s to pgid %d", signal, pgid)
+			err := syscall.Kill(-pgid, signal)
+			if err != nil {
+				log.Infof("Error sending signal to pgid %d: %s", pgid, err)
+			}
 		}
 	}
 	return nil
 }
 
-func sendSignalToProcessGroups(signal syscall.Signal, pgids []int) error {
+// KillTreeWithExcludes sends signal to whole process tree, starting from given pid as root.
+// Omits processes matching names specified in processesToExclude. Kills using pids instead of pgids.
+func KillTreeWithExcludes(signal syscall.Signal, pid int32, processesToExclude []string) error {
+	log.Infof("Will send signal %s to tree starting from %d", signal.String(), pid)
+
+	pgids, err := getProcessGroupsInTree(pid)
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Found process groups: %v", pgids)
+
+	pids, err := findProcessesInGroups(pgids)
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Found processes in groups: %v", pids)
+
+	pids, err = excludeProcesses(pids, processesToExclude)
+	if err != nil {
+		return err
+	}
+
+	signals := wrapWithStopAndCont(signal, pgids)
+	return sendSignalsToProcesses(signals, pids)
+}
+
+func findProcessesInGroups(pgids []int) ([]int, error) {
+	var pids []int
 	for _, pgid := range pgids {
-		log.Infof("Sending signal %s to pgid %d", signal, pgid)
-		err := syscall.Kill(-pgid, signal)
+		cmd := exec.Command("pgrep", "-g", strconv.Itoa(pgid))
+		output, err := cmd.CombinedOutput()
 		if err != nil {
-			log.Infof("Error sending signal to pgid %d: %s", pgid, err)
-			return err
+			return nil, fmt.Errorf("'pgrep -g %d' failed: %s", pgid, err)
```

Comment:
Right, I couldn't find a pure Go implementation so I went for `pgrep`. Do you think adding this to readme is enough? Alternatively I can make an if statement at the beginning of `KillTreeWithExcludes` – if there are no exclusions just go with standard `KillTree`. This way `pgrep` is required only if you define exclusions.

Reply:
IMHO adding this to readme is good enough. The requirement is to have `pgrep` with `-g` option supported. BusyBox pgrep used in Alpine will not work.

Reply:
> 
> 
> Right, I couldn't find a pure Go implementation so I went for `pgrep`. Do you think adding this to readme is enough? Alternatively I can make an if statement at the beginning of `KillTreeWithExcludes` – if there are no exclusions just go with standard `KillTree`. This way `pgrep` is required only if you define exclusions.

I'd rather have it like this, rather than break this part for people who dont use this functionality altogether. Readme comment would just mention that this does not work without `pgrep`

Reply:
How about using https://github.com/mitchellh/go-ps/blob/master/process.go instead of `pgrep`?

Reply:
@janisz I need to find all processes by group ID. I don't see that in the linked API.

Reply:
since Ubuntu is our main target distribution and former Mesosphere generally [supports](https://open.mesosphere.com/downloads/mesos/#apache-mesos-1.9.0) Ubuntu/Debian/CentOs `pgrep` dependency is ok. However package target in Makefile should include `procps` package as dependency.